File: 22-23 Feb/s2.py
def solve():
    n, m = map(int, input().split())
    x=list()
    for i in range(n):
        a,b,c = map(int, input().strip().split())
        x.append((a,b,c))
    x.sort(key = lambda t: t[2])   #根据元组的2号元素排序（编号从0开始）
    for i in range(1, n):
        dx = x[i][0] - x[i-1][0]
        dy = x[i][1] - x[i-1][1]
        dt = x[i][2] - x[i-1][2]
        if dx*dx + dy*dy > dt*dt:
            print(m)
            return
    cnt = 0
    for _ in range(m):
        xx, y, t = map(int, input().split())
        i = 0
        j = n - 1
        while i <= j:            #通过二分查找找到第一个大于(xx,y,t)的元组，这里只需要比较t和x数组中各元素的第2个元素
            mid = (i + j) // 2
            if x[mid][2] < t:
                i = mid + 1
            else:
                j = mid - 1
        flag = False
        if i < n:
            dx = x[i][0] - xx
            dy = x[i][1] - y
            dt = x[i][2] - t
            if dx*dx + dy*dy > dt*dt:
                flag = True
        if i > 0:
            dx = x[i-1][0] - xx
            dy = x[i-1][1] - y
            dt = x[i-1][2] - t
            if dx*dx + dy*dy > dt*dt:
                flag = True
        cnt += flag
    print(cnt)

if __name__ == '__main__':
    T = 1
    for _ in range(T):
        solve()

File: 22-23 Feb/s2.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <tuple>
#define int long long 
#define tiii tuple<int,int,int>

using namespace std;


bool cmp(tiii t1,tiii t2){
    return get<2>(t1)<get<2>(t2);
}

void solve(){
    int n,m;
    cin >> n >> m;
    vector<tiii> x(n);
    for(int i=0;i<n;i++){
        cin>>get<0>(x[i])>>get<1>(x[i])>>get<2>(x[i]);
    }
    sort(x.begin(),x.end(),cmp);
    for(int i=1;i<n;i++){
        int dx = get<0>(x[i])-get<0>(x[i-1]);
        int dy = get<1>(x[i])-get<1>(x[i-1]);
        int dt = get<2>(x[i])-get<2>(x[i-1]);
        if(dx*dx+dy*dy>dt*dt){
            cout<<m;
            return ;
        }
    }
    int cnt=0;
    while(m--){
        int xx,y,t;
        cin>>xx>>y>>t;
        int i = upper_bound(x.begin(),x.end(),tiii(xx,y,t),cmp) - x.begin();
        bool flag = false;
        if(i<n){
            int dx = get<0>(x[i])-xx;
            int dy = get<1>(x[i])-y;
            int dt = get<2>(x[i])-t;
            if(dx*dx+dy*dy>dt*dt){
                flag = true;
            }
        }
        if(i>0){
            int dx = get<0>(x[i-1])-xx;
            int dy = get<1>(x[i-1])-y;
            int dt = get<2>(x[i-1])-t;
            if(dx*dx+dy*dy>dt*dt){
                flag = true;
            }
        }
        cnt+=flag;
    }
    cout<<cnt<<endl;
}

signed main()
{
    int T=1;
    while(T--){
        solve();
    }
    return 0;
}

